### ISSUE TITLE
Rule proposal: no-double-assignments

### ISSUE THREAD

**Issue by denis-sokolov** [eslint/eslint#3481]
Overwriting a variable twice in a row is most likely an error:

```js
cursor_x = 53;
cursor_x = 62;
```

The code above most likely is a result of a copy and paste error and the correct code should be

```js
cursor_x = 53;
cursor_y = 62;
```

See for a comparison [rule V519](http://www.viva64.com/en/d/0108/) in PVS Studio linter for C.

It seems to me to be a fairly useful error detection mechanism, suitable for core.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
@eslint/eslint-team thoughts?

**Comment by gyandeeps** [comment]
>  variable twice in a row
Based on the top statement, is this an error:
```js
x = 1;
y = 2;
x = 3;
```

What about this:
```js
var x = 1;
var y = 2;

x = 23;
```

I am not sure the thought behind this is clear enough and there would be a possibility that this is where a lot of people might have different exceptions. I would say needs more discussion as of now.

**Comment by BYK** [comment]
I think this is reasonable if done in a performant way.

**Comment by denis-sokolov** [comment]
@gyandeeps, neither of those is a warning with the implementation in #3482. It would be great to catch both of these, however, it's not possible to do that without looking behind a lot. Isn't it fine if the rule has false negatives? It doesn't bother anybody. The important part is to not have false positives.

**Comment by mysticatea** [comment]
Do you think about destructuring?

```js
x = 0;
({x, y} = p);
```

**Comment by denis-sokolov** [comment]
@mysticatea, interesting. It would be nice to catch that as well, but the primary source of the error this rule is supposed to catch is copy and paste, though.

**Comment by mysticatea** [comment]
@denis-sokolov Thank you, sounds good enough to me for destructuring. However,

```js
x = 1;
y = 2;
x = 3;
```

I thought this pattern should be caught for the purpose.

**Comment by denis-sokolov** [comment]
@mysticatea, every extra statement to backtrack is a considerable performance hit, though. Just how many statements back would you want to look?

**Comment by mysticatea** [comment]
We can store information while traversing AST.

For example:

```js
var stack;

return {
    "Program": function() { stack = [{}]; },
    "BlockStatement": function() { stack.push({}); },
    "BlockStatement:exit": function() { stack.pop(); }, // and SwitchStatement

    "AssignmentExpression": function(node) {
        var keyMap = stack[stack.length - 1];
        var name = getAssigneeName(node);
        var key = "$" + name; // use prefix `$` to avoid `__proto__`.
        if (name != null) {
            if (keyMap[key] != null) {
                context.report(node, "...");
            } else {
                keyMap[key] = true;
            }
        }
    }
};
```

**Comment by denis-sokolov** [comment]
Ah, @mysticatea, that's a good idea. I will think about this, thank you.

**Comment by gyandeeps** [comment]
@denis-sokolov R u still interested and will be willing to work on this?
@ilyavolodin @nzakas we r doing this right? (Confused as this has accepted and proposal both)

**Comment by denis-sokolov** [comment]
@gyandeeps, this is accepted. That is, if anyone submits an efficient PR, it will be merged.
But my PR was not perfect and it fell off my radar. I should have free time in the holiday season, perhaps I'll clean it up then.

**Comment by alberto** [comment]
@denis-sokolov do you still intend to submit a PR for this?

**Comment by denis-sokolov** [comment]
I suppose it’s safe to close it in the meantime.
